### DISCUSSION TITLE
Extract a string array of union type from Type.Object keys

### DISCUSSION THREAD

**Discussion by dephiros** [sinclairzx81/typebox#405] [Q&A] [answered]
Hi,

I am trying to use `TypeBox` to consolidate some existing API definitions:
```
interface APInterface {
  title: string;
  count: number;
}

const apiAttributes = ['title', 'count'] as const;
```

My attempt:
```
const schema = Type.Object({
  title: Type.String();
  count: Type.Number;
});

type APInterface = Static<typeof schema>; // this works great
const apiAttributes = Object.keys(schema.properties); // the type is now `string[]` instead of a union
```

Would there be some Typescript magic(beside casting `as  keyof APInterface`) or utility that can help make the type of `apiAttributes` a union type?
Took a look at https://github.com/sinclairzx81/typebox/discussions/370 but not sure if applicable. Sorry if I missed another existing question
🙏

**Comment by sinclairzx81** [comment]
@dephiros Hi,

TypeBox supports TypeScript's `keyof` operator with `Type.KeyOf()`. You can use this to extract property keys from the object as well as construct a schema type that matches those keys.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4GUYCGMwAxnAL5wBmUEIcA5AAIDOwAdiQDYHBQD0MdBgBGEAB4MAsAChZQzCjgBePIWIkAPAowQqKAHxxjJ-vzg6lqpMRhcMALjgsYUDgHMcJCAFd2MJ3YfEBEMKApZWW92FysUYQA6AHkRACsMEhgACiQTPLM4aNjkFURZY1t7JzRMBPw3dncsgEocPNNzCuEnBgg0jJgGLHLC339qxIA5YNCoFvb8zrgwWkxYYAwWJwRZcmaFg8PDgrzKx0QLbsYXBvchuABtXFQQiC4smowAIQkEgGkOAATZoAXR69Q8DAowxkRzh8LgJxM3j8AQuOh6QRCYXuTxeYnenx+4n+QNBPWm2KgUPIIwR9MWJnIMIZrOMSKgGAAjj4+BhAU4HowzvcGCj-FCQSy2QykXjXoThMTSexgWDGCl0plpLCZfSCrS5EbLH9SvgiKRtMI9HA-kZ2gUTaUGCK4AAfRjiwaRGRFeCm1Sff4YVBJKgffYHAp+22lHa6vXwpEEdihqiCumJo5I4xIP09EU4DHXVyQnDygkfJW-AGq8mMAAywBgYQIXBpOEzWajS1zoxiaLFY0GRauDBuZcezwVVcwytrap6TZbUDbNK73YdSylG83HRMFbes++NbJ6oYAFV2MAIOwdXvjuZyEA)

```typescript
import { Type, Static } from '@sinclair/typebox'

type T = Static<typeof T>      // type T = { title: string, count: number }

const T = Type.Object({        // const T = {
  title: Type.String(),        //   type: 'object',
  count: Type.Number()         //   properties: {
})                             //     title: { type: 'string', [Symbol(TypeBox.Kind)]: 'String' },
                               //     count: { type: 'number', [Symbol(TypeBox.Kind)]: 'Number' }
                               //   },
                               //   required: [ 'title', 'count' ],
                               //   [Symbol(TypeBox.Kind)]: 'Object'
                               // }


type K = Static<typeof K>      // type K = 'title' | 'count'

const K = Type.KeyOf(T)        // const K = {
                               //   anyOf: [
                               //     { const: 'title', type: 'string', [Symbol(TypeBox.Kind)]: 'Literal' }, 
                               //     { const: 'count', type: 'string', [Symbol(TypeBox.Kind)]: 'Literal' }
                               //   ],
                               //   [Symbol(TypeBox.Kind)]: 'Union'
                               // }
```
If you only need the keys as a type (with no schema), you can use the following which avoids the need to prop dive through `T.properties`.

```typescript
type K = keyof Static<typeof T> // just the static type
```
Hope this helps!
S

**Comment by dephiros** [comment]
Hi @sinclairzx81 ,

Thank you for the answer and it pointed me to the right direction. What I wanted is an actual array of properties key but with the union type.

This is a bit clunky but produces the right result:
```
const apiAttributes  = Type.KeyOf(schema).anyOf.map((x) => x.const)
```

Though for some reasons, when I tried to extract to a helper method, Typescript is not quite happy:
```
function keyOf<P extends TProperties = {}>(schema: TObject<P>) {
  // x cannot be infer so it becomes `any`
  return Type.KeyOf(schema).anyOf.map((x) => x.const)
}
```

**Reply by sinclairzx81** [reply, marked as answer]
@dephiros Hi, you can try the following.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/LAKAlgtgDg9gTgFwAQG8kBUCeUCmAaJAZQQEMEwBjA9QigCxwhIIFUA7MGN9GdAVygAbfBmw4A4nxJwAJkgC+SAGZwYEJAHIAAgGcwbCoJJg4AegRiARjAAeG0KFOmkAWjfuPnr95++--l0dnAHlLACscCgQAaRxMHSDXAOSU1IDQHBtYRCQLXAxiOH0Ac1j4gB50JEyEHDYZHQL6RhIAPiQAXiR2Tm5eAWFygGs4mCUiUnIKStbWhxBM7OQlPgNyLgmitlK4nUrqm1r6xppmplaACh0zkgAuDABKe5oELZ2K9HaUUCQkCi4dMgRphgkpnrQGExOqJcAA6Mqgq43B4-P4AoG7aFYXCSaQyWHoHpcC7A0EPX5IAD8SFJSlhJDYILpTCgF0EYFqcBIgk67XZnO5sP+bEB5NuqIpkqlMIkUlkBIAMhycFzBCS4mSqUgANq0oXogC6SHu2oNqLgOAQfDgbBpmJIJ0KJTKe0+oHk8ycSTSPt9KUSLB0JGKOESfvDEe8oGFgKQAEEsWJYaEIlELmgbM8kwA5PgQSwqi4PAiYLNw3P5wvk+QokAx5AAIUTcJTkQQ6aQAC8yzhYRWC3Aiwpa-WkABJZu9sdsTk6NsXbVxggNg216PopDRACM0Kd2xdFzH5K92o0dgIGkwGgvnY0ZrrG+iACZd69nbsLnHj85T+fNFf71HaIAGZXzeA8G2-HUNFvA0gA)

```typescript
import { Type, Static, TSchema, UnionToTuple, TypeGuard } from '@sinclair/typebox'

// -------------------------------------------------------------------
// ObjectKeys
// -------------------------------------------------------------------
export type TStringKeys<T extends TSchema> = UnionToTuple<keyof Static<T>>

export function StringKeys<T extends TSchema>(schema: T): TStringKeys<T> {
  const keyOf: TSchema = Type.KeyOf(schema)
  const keys = TypeGuard.TUnion(keyOf)   ? keyOf.anyOf.map(literal => literal.const) :
               TypeGuard.TLiteral(keyOf) ? [keyOf.const] : []
  return keys as TStringKeys<T>
}

// -------------------------------------------------------------------
// Usage
// -------------------------------------------------------------------
const A = Type.Object({ x: Type.Number(), y: Type.Number() })
const B = Type.Object({ z: Type.Number() })
const I = Type.Intersect([A, B])

const K1 = StringKeys(I) // ['x', 'y', 'z']
const K2 = StringKeys(A) // ['x', 'y']
const K3 = StringKeys(B) // ['z']
```

**Reply by dephiros** [reply]
That works great! Thank you 🙏

**Reply by sinclairzx81** [reply]
@dephiros Hey, no problem. I've actually made a quick update to this example to support generalized types of `TSchema`. I'd recommend this implementation over the previous `TObject` one as it generalizes to `TUnion` and `TIntersect`.
